homeassistant/components/camera/logi.py

```diff
@@ -0,0 +1,204 @@
+
+    def __init__(self, hass, camera, device_info):
+        """Initialize Logi camera."""
+        super(LogiCam, self).__init__()
+        self._camera = camera
+        self._name = self._camera.name
+
+    @property
+    def name(self):
+        """Return the name of this camera."""
+        return self._name
+
+    @property
+    def supported_features(self):
+        """Logi Circle camera's support turning on and off ("soft" switch)."""
+        return SUPPORT_ON_OFF
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        return {
+            ATTR_ATTRIBUTION: CONF_ATTRIBUTION,
+            'battery_saving_mode': ('on' if self._camera.battery_saving
+                                    else 'off'),
+            'firmware': self._camera.firmware,
+            'generation': self._camera.model_generation,
+            'ip_address': self._camera.ip_address,
+            'mac_address': self._camera.mac_address,
+            'microphone_gain': self._camera.microphone_gain,
+            'model': self._camera.model,
+            'mount': self._camera.mount,
+            'plan': self._camera.plan_name,
+            'speaker_volume': self._camera.speaker_volume,
+            'timezone': self._camera.timezone
+        }
+
+    async def async_camera_image(self):
+        """Return a still image from the camera."""
+        return await self._camera.get_snapshot_image()
+
+    async def async_turn_off(self):
+        """Disable streaming mode for this camera."""
+        await self._camera.set_streaming_mode(False)
+
+    async def async_turn_on(self):
+        """Enable streaming mode for this camera."""
+        await self._camera.set_streaming_mode(True)
+
+    @property
+    def should_poll(self):
+        """Update the image periodically."""
+        return True
+
+    async def set_config(self, mode, value):
+        """Set an configuration property for the target camera."""
+        if mode == LED_MODE_KEY:
+            await self._camera.set_led(value)
+        if mode == PRIVACY_MODE_KEY:
+            await self._camera.set_privacy_mode(value)
+        if mode == BATTERY_SAVING_MODE_KEY:
+            await self._camera.set_battery_saving_mode(value)
+
+    async def download_livestream(self, filename, duration):
+        """Download a recording from the camera's livestream."""
+        # Render filename from template.
+        filename.hass = self.hass
+        stream_file = filename.async_render(
+            variables={ATTR_ENTITY_ID: self.entity_id})
+
+        # Respect configured path whitelist.
+        if not self.hass.config.is_allowed_path(stream_file):
+            _LOGGER.error(
+                "Can't write %s, no access to path!", stream_file)
+            return
+
+        asyncio.shield(self._camera.record_livestream(
+            stream_file, timedelta(seconds=duration)), loop=self.hass.loop)
+
+    async def livestream_snapshot(self, filename):
```

Comment:
A quick explanation on why this was implemented when there's already a generic `snapshot` service for cameras:

When not live streaming, Logi cameras (particularly battery-powered devices) go into a low power state, only taking a snapshot every 1-5 minutes. This cached image is what the `async_camera_image` method uses.

This method exists if you _really_ need to wake the camera from sleep immediately to take a snapshot. I didn't want this to be the default source for camera stills as it increases battery drain.